### PR TITLE
Clean inbound authentication metadata under RCS

### PR DESCRIPTION
**Pull request by n1v0lg**
With RCS, authentication metadata received as part of cross cluster requests can include arbitrary, user-supplied fields. This PR adds logic to clean inbound authentication metadata, leaving only minimally-required internal fields.

### CODE REVIEW

**Comment by n1v0lg** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 1]
```diff
@@ -513,8 +569,7 @@ public void writeTo(StreamOutput out) throws IOException {
         doWriteTo(effectiveSubject, authenticatingSubject, type, out);
     }
 
-    // Package private for testing
-    static void doWriteTo(Subject effectiveSubject, Subject authenticatingSubject, AuthenticationType type, StreamOutput out)
+    private static void doWriteTo(Subject effectiveSubject, Subject authenticatingSubject, AuthenticationType type, StreamOutput out)
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:572]: Noticed this wasn't used anywhere, so switching it back to private.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 2]
```diff
@@ -810,7 +865,7 @@ private void checkConsistencyForApiKeyAuthenticationType() {
                         + "a non-null serialized cross cluster access role descriptors field"
                 );
             }
-            checkNoRunAs(this, "Remote access");
+            checkNoRunAs(this, "Cross cluster access");
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:868]: Unrelated nit fix

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/RoleDescriptorStore.java [hunk 3]
```diff
@@ -139,7 +139,7 @@ public void resolveCrossClusterAccessRoleReference(
     ) {
         final Set<RoleDescriptor> roleDescriptors = crossClusterAccessRoleReference.getRoleDescriptorsBytes().toRoleDescriptors();
         for (RoleDescriptor roleDescriptor : roleDescriptors) {
-            if (hasPrivilegesOtherThanIndex(roleDescriptor)) {
+            if (roleDescriptor.hasPrivilegesOtherThanIndex()) {
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/RoleDescriptorStore.java:142]: Sorry, I messed up addressing a [comment](https://github.com/elastic/elasticsearch/pull/94968#discussion_r1157897658) in the previous PR. Addressing it here.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 4]
```diff
@@ -282,6 +284,61 @@ private static Map<String, Object> maybeRewriteMetadata(TransportVersion olderVe
         }
     }
 
+    /**
+     * Creates a copy of this Authentication instance, but only with metadata entries specified by `fieldsToKeep`.
+     * All other entries are removed from the copy's metadata.
+     */
+    public Authentication copyWithFilteredMetadataFields(final Set<String> fieldsToKeep) {
+        Objects.requireNonNull(fieldsToKeep);
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:292]: I decided to keep this generic, and non-restrictive. For instance, the method does not enforce that `fieldsToKeep` contains the API key ID metadata field for an API-key authentication. If fields are filtered out that "break" the authentication instance, this is caught by the internal consistency checks further down. It feels cleaner to rely on those, rather than attempt to duplicate them here.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java [hunk 5]
```diff
@@ -282,6 +284,61 @@ private static Map<String, Object> maybeRewriteMetadata(TransportVersion olderVe
...
+        if (fieldsToKeep.isEmpty()) {
+            return copyWithEmptyMetadata();
+        }
+        final Map<String, Object> metadataCopy = new HashMap<>(getAuthenticatingSubject().getMetadata());
+        final boolean metadataChanged = metadataCopy.keySet().retainAll(fieldsToKeep);
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java:297]: Decided against an idempotency check (i.e., don't copy if it's unnecessary) here to keep the contract of the method clear. Callers can rely on always getting a new instance this way and don't have to worry about side-effects on the original instance. The idempotency check would only avoid the creation of a fairly small object, so not much performance gain.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java [hunk 6]
```diff
@@ -69,6 +89,15 @@ public Authentication getAuthentication() {
         return authentication;
     }
 
+    public CrossClusterAccessSubjectInfo cleanWithValidation() {
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java:92]: I've moved the validation logic that used to sit in the service class here for a cleaner encapsulation of cleaning and validating an inbound subject info instance. This, for example, allows us to first check the subject type, then clean, then validate the clean version without exposing all these internal steps and their dependencies beyond this class. In the end, all that callers will see is that they got a clean and valid version or a validation error.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java [hunk 7]
```diff
@@ -37,7 +42,22 @@
 import java.util.Set;
 
 public final class CrossClusterAccessSubjectInfo {
+
     public static final String CROSS_CLUSTER_ACCESS_SUBJECT_INFO_HEADER_KEY = "_cross_cluster_access_subject_info";
+    private static final Logger logger = LogManager.getLogger(CrossClusterAccessSubjectInfo.class);
+    private static final Set<String> API_KEY_AUTHENTICATION_METADATA_TO_KEEP = Set.of(
+        AuthenticationField.API_KEY_ID_KEY,
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java:49]: I've included these here since they cross-cluster-access specific: for example, the fact that we keep some of the API-key metadata fields but not other only makes sense in the RCS context.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java [hunk 8]
```diff
@@ -69,6 +89,15 @@ public Authentication getAuthentication() {
         return authentication;
     }
 
+    public CrossClusterAccessSubjectInfo cleanWithValidation() {
+        // Need to do this first. Otherwise, the `copyWithFilteredMetadataFields` call in `copyAuthenticationWithCleanMetadata` may fail
+        // with a confusing error message for unsupported types
+        ensureAuthenticationHasSupportedSubjectType();
+        final var cleanCopy = new CrossClusterAccessSubjectInfo(copyAuthenticationWithCleanMetadata(), roleDescriptorsBytesList);
+        cleanCopy.validate();
+        return cleanCopy;
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java:98]: Why cannot we call `validate()` first? Validate internally check type first and does not seem to rely on the clean copy. 
We also might not need assert the type in `copyAuthenticationWithCleanMetadata` since as is it is checked again in `validate`. If we move `validate` to be the first call. We can just assert the type here afer the cleanCopy if we really want to.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java [hunk 9]
```diff
@@ -923,6 +926,31 @@ public void testMaybeRewriteMetadataForCrossClusterAccessAuthentication() throws
...
+            10,
+            () -> new Tuple<>(ESTestCase.randomAlphaOfLength(20), ESTestCase.randomAlphaOfLength(10))
+        );
+        final Map<String, Object> allMetadata = new HashMap<>(metadataToKeep);
+        allMetadata.putAll(
+            ESTestCase.randomMap(0, 10, () -> new Tuple<>(ESTestCase.randomAlphaOfLength(30), ESTestCase.randomAlphaOfLength(10)))
+        );
+        // Using service account since it allows us to easily specific metadata, but will not fail instantiation if any service-account
+        // specific metadata is missing
+        final Authentication authentication = AuthenticationTestHelper.builder().serviceAccount().metadata(Map.copyOf(allMetadata)).build();
```
Comment (ywangd) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java:941]: Not like it is important, but the fact we don't check service-account specific metadata could be a separate bug. No need to do anything here.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/CrossClusterAccessAuthenticationService.java [hunk 10]
```diff
@@ -98,9 +95,8 @@ public void authenticate(final String action, final TransportRequest request, fi
                     // try-catch so any failure here is wrapped by `withRequestProcessingFailure`, whereas `authenticate` failures are not
                     // we should _not_ wrap `authenticate` failures since this produces duplicate audit events
                     try {
-                        final CrossClusterAccessSubjectInfo crossClusterAccessSubjectInfo = crossClusterAccessHeaders.subjectInfo();
-                        validate(crossClusterAccessSubjectInfo);
-                        writeAuthToContext(authcContext, authentication.toCrossClusterAccess(crossClusterAccessSubjectInfo), listener);
+                        final CrossClusterAccessSubjectInfo subjectInfo = crossClusterAccessHeaders.subjectInfo().cleanWithValidation();
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/CrossClusterAccessAuthenticationService.java:98]: I wonder whether `crossClusterAccessHeaders` should not expose the subjectInfo that has not been cleaned and validated at all? That is, can we have something like `crossClusterAccessHeaders.getCleanAndValidatedSubjectInfo()` which internally calls `subjectInfo.cleanWithValidation()` and remove `crossClusterAccessHeaders.subjectInfo()`?

**Review by ywangd — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java [hunk 11]
```diff
@@ -151,6 +180,56 @@ public Map<String, Object> copyWithCrossClusterAccessEntries(final Map<String, O
...
+        } else {
+            return authentication.copyWithEmptyMetadata();
+        }
+    }
+
+    private boolean authenticationHasSupportedSubjectType() {
+        final Subject effectiveSubject = authentication.getEffectiveSubject();
+        return effectiveSubject.getType().equals(Subject.Type.USER)
+            || effectiveSubject.getType().equals(Subject.Type.SERVICE_ACCOUNT)
+            || effectiveSubject.getType().equals(Subject.Type.API_KEY);
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java:211]: Good call. I'll tweak.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java (hunk 8 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java:98]: Mostly because I want to make sure we validate the object we actually use in the remaining flow. It seems a little counter intuitive to validate, then change the object, then assume it's still valid. If our validation logic evolves to include more metadata checks we fail when we shouldn't (because we fail on something being present that would get cleaned) and not fail when we should (because we cleaned too much). Ideally, we wouldn't even check the type first but that leads to confusing error messages, since `copyAuthenticationWithCleanMetadata` doesn't always produce valid Authentication instance for unsupported types.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java (hunk 9 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java:941]: Ya agreed, we probably want to enforce this as a consistency check but I didn't look too closely.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java (hunk 8 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java]: LMK if you feel strongly, otherwise I'll keep (largely) as is. I've pushed a change to inlines the type check and asserts `!authentication.isCrossCluster()` in other places since I've adopted your deny-list suggestion.

Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/CrossClusterAccessSubjectInfo.java]: No strong feeling at all. Thanks for asking.

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch
